BUG: Add basic __format__ for masked element to fix incorrect print

5tefan opened the pull request:
Fix and test conversion of masked element to string in fstring / string interpolation. See PEP 498 for string interpolation (aka fstrings) in Python 3.6 and above.

Fixes #15409.

5tefan committed:
TST: adds fstring test for np.ma.masked element

5tefan committed:
BUG: fix fstring conversion for masked element

seberg commented on numpy/ma/core.py:
```diff
@@ -6397,6 +6397,9 @@ def __repr__(self):
             return object.__repr__(self)
 
+    def __format__(self, format_spec):
+        return str(masked_print_option._display)
```
Should this reject some format spec's (or the other way around, only accept certain ones)?

5tefan replied:
Hmm. That's a good question. I'll dig around and see if I can figure out how, perhaps nan is treated, as a model for this. Not sure I'm in a position to implement the full https://www.python.org/dev/peps/pep-3101/#standard-format-specifiers spec.

seberg replied:
Well, we could start off a bit limited possibly. Maybe some of it can be handled by formatting twice (not quite sure). But if the user requests a certain alignment, it would not be nice to just ignore that either.

eric-wieser replied:
One option would be `format(str(masked_print_option._display), format_spec)`

seberg replied:
Yeah, I think I like that option. There is a bit of a difficulty around whether or not to support `f` or `d`, but basically going the string way is a good safe bet.

Only issue is that the error would read:
```
ValueError: Unknown format code 'd' for object of type 'str'
```
I suppose we could replace that with our own error.

5tefan replied:
Great suggestion, that looks correct and (I missed earlier) what's suggested in the PEP https://www.python.org/dev/peps/pep-3101/#controlling-formatting-on-a-per-type-basis

5tefan replied:
Actually, this seems important since this will be used in a context where users are probably expecting a integer or floating type. Thoughts on slicing last char from `format_string` if it ends in a `d` or `f`?

eric-wieser replied:
That's a good point. Perhaps we should not open this can of worms in this PR, and just require for now that the spec is the empty string.

There's an open issue about implementing `__format__` in numpy.

5tefan replied:
Hmm, unfortunately, it seems partway open now... But to make some tradoffs explicit: 

- Doing nothing else now gives misleading error message (user thought they were formatting a float)
- Empty format string sacrifices ability to align (check if empty, or impose empty?)
- Introduce conditional to slice format type, but may have performance impact (how important is performance for string formatting?)

However, I think any of these cases are better than the current behaviour printing "0.0".

5tefan replied:
And a fourth option, I believe what @seberg was suggesting,

- except and raise a better exception. Perhaps something like "Masked arr marker is str type, omit presentation type when formatting masked array"

seberg replied:
Yeah, although getting the error details right is maybe not trivial. I suppose you could chain exceptions, but I am not sure it actually helps a lot, I sometimes feel the typical scientist "new to programming" user already needs persuasion to parse through a single traceback...

seberg commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,13 @@ def test_deepcopy(self):
+        # Seeing that as of 20200123, the atravis do not target versions under 3.6, so this is ok.
+        a = array([0, 1, 2], mask=[False, True, False])
+        assert_equal(f"{a}", "[0 -- 2]")
+        assert_equal(f"{masked}", "--")
```
Mindblowing, I did not realize we can now rely on the availability of f-strings ;).

charris replied:
Since we dropped Python 3.5. There are a *lot* of places where they would be a style improvement.

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,13 @@ def test_deepcopy(self):
         assert_equal(copied.mask, [0, 0, 0])
         assert_equal(a.mask, [0, 1, 0])
 
+    def test_fstring(self):
```
I think perhaps better to just call this `test_format`? f-strings are a distraction here, what you're actually testing is the `format()` builtin

5tefan replied:
Ahh, this is a good point. i agree. Thanks!

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,13 @@ def test_deepcopy(self):
+        # fstrings (PEP 498) are a SyntaxError in versions under Py 3.6.
+        # Seeing that as of 20200123, the atravis do not target versions under 3.6, so this is ok.
+        a = array([0, 1, 2], mask=[False, True, False])
+        assert_equal(f"{a}", "[0 -- 2]")
```
```suggestion
        assert_equal(format(a), "[0 -- 2]")
```

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,13 @@ def test_deepcopy(self):
+        # Seeing that as of 20200123, the atravis do not target versions under 3.6, so this is ok.
+        a = array([0, 1, 2], mask=[False, True, False])
+        assert_equal(f"{a}", "[0 -- 2]")
+        assert_equal(f"{masked}", "--")
```
Even though they are supported, I think we should be testing this at a more primitive level:
```suggestion
        assert_equal(format(masked), "--")
```

5tefan committed:
BUG, TST: improve masked __format__ and tests.

Thanks @eric-wieser and @seberg for review and suggestions!

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,16 @@ def test_deepcopy(self):
 
+    def test_format(self):
+        # Note: Python 3.6 and above fstring uses __format__, see PEP 498.
+        # Testing with underlying primitive.
```
Fyi, `"{: >5}".format(a)` also uses `format`, f-strings are not a new mechanism.

seberg replied:
I think we can just drop the whole comment, I do not think it really adds anything.

eric-wieser replied:
Agreed, I think dropping this comment is sensible

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -6397,6 +6397,9 @@ def __repr__(self):
             return object.__repr__(self)
 
+    def __format__(self, format_spec):
+        return format(str(masked_print_option._display), format_spec)
```
```suggestion
        # Replace ndarray.__format__ with the default, which supports no format characters.
        # Supporting format characters is unwise here, because we do not know what type
        # the user was expecting - better to not guess.
        return object.__format__(self, format_spec)
```

5tefan committed:
BUG: fix ma format, delegate to object.__format__

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,16 @@ def test_deepcopy(self):
+        assert_equal(format(masked), "--")
+        assert_equal(format(masked, ""), "--")
+        assert_equal(format(masked, " >5"), "   --")
+        assert_equal(format(masked, " <5"), "--   ")
```
You'll need to remove these tests

seberg replied:
Ah ok, I do not mind not supporting them for now. So removal or error check with a comment that this could be changed in the future sounds good to me.

eric-wieser replied:
Yeah, adding support for formatting is a big task I do not think we should be attempting to address in this bugfix PR.

seberg reviewed (commented):
Should we support the `str` format characters as before? I guess those are the reason for the test failures?

5tefan committed:
BUG: incorp review re. testing ma format.

seberg commented:
OK, thanks looks good to me as well. Followup thoughts appreciated, although I guess they may be pretty complex...

seberg commented:
Hmmm, holding off for one second: This will mean that code that currently prints nonsense could instead crash. I am leaning towards that probably being fine here since it should not be common code, but we could deprecate/warn in principle...

eric-wieser commented:
So perhaps instead:
```
try:
    return object.__format__(self, format_string)
except Exception:
    warnings.warn(
        "Format strings passed to MaskedConstant are ignored, but in future may error or produce different behavior",
        FutureWarning
    )
    return object.__format__(self, "")
```

seberg commented:
Or `super().__format__(self, format_string)` which is 100% the old behaviour?

eric-wieser commented:
Right, but code that fails because it expected `0.0` in the format string but got `--` probably deserves to fail

seberg commented:
True, so that way still something bogus gets printed (leading to no crash), but a better bogus value. I am drifting around +-0 for all of the options, am even OK with just switching if it seems too ugly.

seberg commented:
@5tefan would you be up to looking at Eric's suggestion and adding a test for the deprecated versions? There is some complexity around it, but I think I would be happier with that version, and having a deprecation warning for a bit doesn't really hurt.

5tefan committed:
BUG: ensure backwards compatibility printing MaskedConstant

Concerns raised to avoid causing old code to blow up. Introduces warning to
avoid breaking anything that previously printed "0.0" using a format string.
Now, FutureWarning given and format string is ignored.

5tefan commented:
Alright, probably a good idea. I suspect you're correct that it's uncommon, but just in case. Looks like this is consistent with the way warnings are tested elsewhere there.

seberg commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,22 @@ def test_deepcopy(self):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.filterwarnings("always")
+            assert_equal(format(masked, " >5"), "--")
+            assert_equal(len(w), 1, "Expected a FutureWarning for using format_spec with MaskedElement")
```
Thanks, can you write this test with a simple `with assert_warns(...):` that should read quite a bit nicer. Otherwise I think we should just go ahead with this.

5tefan replied:
Oh, I see. The construct I used is actually used to test absence of warnings.

5tefan committed:
TST: use assert_warns instead of catch_warnings

5tefan commented:
@seberg, is this ready to go? Hoping it hasn't fallen by the wayside.

seberg commented on numpy/ma/core.py:
```diff
@@ -6397,6 +6397,20 @@ def __repr__(self):
+        # the user was expecting - better to not guess.
+        try:
+            return object.__format__(self, format_spec)
+        except Exception:
```
I think we know that `ValueError` is expected, so that would be cleaner (e.g. the super unlikely case that someone hits `ctrl+c` at exactly this time.

WarrenWeckesser replied:
Looks like we a get a `TypeError`, not a `ValueError`, e.g.:

```
In [23]: object.__format__(123, '<5')                                                                                                                                       
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-8ccbea622452> in <module>
----> 1 object.__format__(123, '<5')

TypeError: unsupported format string passed to int.__format__
```

seberg commented on numpy/ma/tests/test_core.py:
```diff
@@ -447,6 +447,20 @@ def test_deepcopy(self):
+        # assert_equal(format(masked, " <5"), "--   ")
+
+        # Expect a FutureWarning for using format_spec with MaskedElement
+        with_format_string = assert_warns(FutureWarning, format, masked, " >5")
```
Would be nicer as a with statement:
```
with assert_warns(FutureWarning):
    ...
```

seberg reviewed (approved):
Yes, sorry. I had really mainly waited for the test IIRC, two nitpicks although if you are tired of this, I can fix it up.

WarrenWeckesser commented:
@5tefan, there are two remaining changes suggested by @seberg.  Will you be able to fix those?

WarrenWeckesser commented:
Thanks @5tefan, @eric-wieser and @seberg.

The changes in this PR were included in https://github.com/numpy/numpy/pull/15794, so closing.
